crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -8,126 +8,26 @@ use std::collections::HashMap;
-                                error!("Failed to commit write batch while compacting objects");
-                                wb = objects.batch();
-                                num_pending_wb_ops = 0;
-                                break;
-                            } else {
-                                info!("Committed write batch while compacting objects, keys scanned = {:?}, objects scanned = {:?}", total_keys_scanned, total_objects_scanned);
-                            }
-                            wb = objects.batch();
-                            num_pending_wb_ops = 0;
-                        }
-                    }
-                    num_versions_for_key = 0;
-                    object_id = Some(key.0);
-                    delete_range_end = None;
-                }
-                num_versions_for_key += 1;
-                // We'll keep maximum `num_versions_to_retain` latest version of any object
-                delete_range_start = match delete_range_end {
-                    Some(_end) => {
-                        total_pruned += 1;
-                        assert!(num_versions_for_key > num_versions_to_retain);
-                        Some(key.1)
-                    }
-                    None => {
-                        if num_versions_for_key == num_versions_to_retain {
-                            delete_range_end = Some(key.1)
-                        }
-                        None
-                    }
-                };
-            } else {
-                num_versions_for_key = 1;
-                total_objects_scanned = 1;
-                object_id = Some(key.0);
-            }
-        }
-        if let Some(obj_id) = object_id {
-            if let (Some(start_seq_num), Some(end_seq_num)) = (delete_range_start, delete_range_end)
-            {
-                let start = ObjectKey(obj_id, start_seq_num);
-                let end = ObjectKey(obj_id, end_seq_num);
-                if let Ok(new_wb) = wb.delete_range(objects, &start, &end) {
-                    wb = new_wb;
-                    num_pending_wb_ops += 1;
-                } else {
-                    error!("Failed to invoke delete_range on write batch while compacting objects");
-                    wb = objects.batch();
-                    num_pending_wb_ops = 0;
-                }
-            }
-        }
-        if num_pending_wb_ops > 0 {
-            if wb.write().is_err() {
-                error!("Failed to commit write batch while compacting objects");
-            } else {
-                info!("Committed write batch while compacting objects, keys scanned = {:?}, objects scanned = {:?}", total_keys_scanned, total_objects_scanned);
-            }
-        }
-        info!(
-            "Finished compacting objects, keys scanned = {:?}, objects scanned = {:?}",
-            total_keys_scanned, total_objects_scanned
-        );
-        total_pruned
-    }
-
     fn handle_checkpoint(
```

Comment:
I know this is not a place where this PR modifies things but it should:
* Again just a small doc to tell us what this fn should do, `handle_checkpoint` really means too little.
* The `skip_prior_to` operation is expensive. It is safe to use as the first key (object_id, 0) and last key (object_id, version.inc()).
* The use of `delete_range` is a problem, since it call `delete_range_cf`, and the docs specifically say you should not do so for many small ranges (see: https://docs.rs/rocks/latest/rocks/db/struct.DB.html#method.delete_range_cf).
* Using an iterator over the DB as in `iter.take_while` purely to compute `pruned` which I suspect is just a metric is a pretty expensive thing to do. (Which is then ignored in many places.)

Reply:
> handle_checkpoint really means too little

agree, will somewhat change naming
>  delete_range concern
rocksdb stores delete_range intervals in the separate DS. Multiple delete_ranges might indeed be an issue. We observed it in a past. For example, during `get` operation rocksdb will typically try to apply and merge them to return the result. That's why we're currently passing `ignore_range_deletions` flag to objects. So now deleted ranges are stored separately and completely ignored by reads. Basically they are only visible for compaction job to permanently remove the data
We ran experiment on private-testnet with current PR and we don't see any issues.
For context, as a followup, we're still trying to migrate from delete_range to point-delete API. It will somewhat simplify this function and will allow to enable transactions on given rocksdb instance. The concern is that point-deletes in rocksdb are creating tombstones, thus iterator seek performance might be impacted. We're running experiments now to validate that
> take_while concern
it's used not for a metric, but to form a corresponding delete range. But it really might be unnecessary